src/storage/blocksstable/ob_filter_cache.cpp

```diff
@@ -329,12 +383,85 @@ DEFINE_DESERIALIZE(ObBloomFilterCacheValue)
   return ret;
 }
 
-DEFINE_GET_SERIALIZE_SIZE(ObBloomFilterCacheValue)
+DEFINE_GET_SERIALIZE_SIZE(ObFilterCacheValue)
 {
-  return bloom_filter_.get_serialize_size() + serialization::encoded_length_i16(version_) +
+  return filter_->get_serialize_size() + serialization::encoded_length_i16(version_) +
          serialization::encoded_length_i16(rowkey_column_cnt_) + serialization::encoded_length_vi32(row_count_);
 }
 
+/**
+ * ----------------------------------------------------ObFilterBuilder----------------------------------------------------
+ */
+
+ObFilterBuilder::ObFilterBuilder()
+    : allocator_(), size_(0), is_xor_(false), inited_(false), rowkeys_(NULL), filter_value_(NULL)
+{}
+
+ObFilterBuilder::~ObFilterBuilder()
+{
+  allocator_.reset();
+  rowkeys_ = NULL;
+  filter_value_ = NULL;
+}
+
+int ObFilterBuilder::init(const int64_t rowkey_column_cnt, const int64_t size, const bool is_xor, ObFilterCacheValue* filter_value)
+{
+  int ret = OB_SUCCESS;
+
+  size_ = size;
+  is_xor_ = is_xor;
+  filter_value_ = filter_value;
+  key_pos_ = 0;
+  char* buf;
+
+  if (OB_FAIL(filter_value_->init(rowkey_column_cnt, size, is_xor))) {
+    STORAGE_LOG(WARN, "fail to init filter cache value", K(ret), KP(filter_value_));
+  } else if (NULL == (buf = (char*)allocator_.alloc(sizeof(ObStoreRowkey) * size))) {
+    ret = OB_ALLOCATE_MEMORY_FAILED;
+    STORAGE_LOG(ERROR, "fail to allocate memory, ", K(ret));
+  } else {
+    rowkeys_ = new (buf) ObStoreRowkey[size_];
+    inited_ = true;
+  }
+  return ret;
+}
+
+int ObFilterBuilder::insert(ObStoreRowkey& rowkey)
+{
+  int ret = OB_SUCCESS;
+
+  if (is_xor_) {
+    ObIAllocator* allocator = &allocator_;
+    if (key_pos_ >= size_) {
+      ret = OB_SIZE_OVERFLOW;
+      STORAGE_LOG(WARN, "rowkeys is already full", K(ret));
+    } else if (OB_FAIL(rowkey.deep_copy(rowkeys_[key_pos_], *allocator))) {
```

Comment:
we can also support insert single rowkey for xor filter builder, then the big space to keep all the rowkeys is no longer needed